### DISCUSSION TITLE
Question on restreaming ONVIF (and my experience with Wyze Cams)

### DISCUSSION THREAD

**Discussion by rocket59** [blakeblackshear/frigate#9856] [Ask A Question] [answered]
I plan on getting a Reolink doorbell which I understand sends the button press as part of the onvif stream. As frigate doesn't support the button, I plan on using Scrypted or Home Assistant to do the button detection and use frigate for recording and object detection. So my question is: if we restream the camera from frigate to scrypted/HA is this data preserved ? If so do we use the standard parameters as per the documentation? 

Also sharing my experience with Wyze cams (v2/v3). I was previously accessing 2 RTSP streams from the cameras for different purposes with other software. And since I started I kept getting constant ffmpeg crashes/detection issues which are generally understood to be primarily due to bad connections/flaky wifi. My aha moment was to restream one of the troublesome cameras and interestingly most of the random errors on the others went away! The system is now much more stable (more so after using restream on the rest of the cameras. In hindsight this was a pretty obvious conclusion (those low powered CPUs really struggle to provide more than 1 stream) but I was considering replacing them with the more reliable dahuas/hikvisions.

**Comment by NickM-27** [comment]
> if we restream the camera from frigate to scrypted/HA is this data preserved ? 

The camera stream and doorbell press are completely separate pieces of data. They are not packaged together.

**Reply by rocket59** [reply]
Thanks - so it's not possible to pass this through from frigate?

**Reply by NickM-27** [reply, marked as answer]
There's nothing to pass through, the HA integration works using the reolink API directly so it must communicate with the doorbell directly

**Reply by rocket59** [reply]
Got it. Thanks.

**Reply by mayo** [reply]
@NickM-27 I've started poking at this couple of days ago and looking at the debug output from the Reolink HA integration. I'm pretty certain they use ONVIF to receive the button event.

I know this is currently not supported in Frigate (at least it doesn't seem to be), but is there interest in having this kind of out of band actions exposed in Frigate? I'd be interested in having a go at it, if it's something that will be welcome. Does anyone know if there are folks working on anything like this already?

**Reply by NickM-27** [reply]
It does not use onvif only, we can see this because the reolink doorbell must be able to connect to HomeAssistant via http across the same local network. (see the issues regarding self signed certs and using across vlans).

In any case, the number of doorbells that support ONVIF and sending this event seems to be 1, I do not believe that this is worth the effort of maintaining or supporting this as it will just take time out of supporting features that will benefit more users. 

Other maintainers may have a different opinion

**Reply by mayo** [reply]
I'm pretty certain there are other doorbells that do this (for eg. Amcrest AD410), as well as other cameras push values that can be useful such as sound detection, tamper sensors, storage errors, etc. 

Thank you for the honest response, though, and saving my time. I'll find other ways to integrate these.

**Reply by NickM-27** [reply]
I used to have the ad410 before I got the reolink doorbell. Doorbell notifications on the amcrest required their http API as well

**Reply by mayo** [reply]
I'm not sure what you mean by "their" here. In both cases the button presses and all the other events I mentioned above are exposed via the onvif event handling, which is part of the core spec, and not a custom vendor API as is implied in your message above. (See https://www.onvif.org/ver10/events/wsdl/event.wsdl)

Yes, it relies on a http connection from the device to the client, so the device can call back an endpoint, which may suffer with self signed certs and/or other TLS issues, but as long as an error can be presented, it is exercise for the user to deal with. (I'm not entirely sure what the issues with VLAN are, since frigate already has to be able to receive packets from cameras and vice versa, so there must already be a path between the two.) Subscribing and receiving some events is also possible via the (RTP) stream and it comes in the metadata portion, so HTTP connection can be potentially cut out completely (though at the moment I'm not 100% sure if this specific event I'm looking for can be included in the event stream, but I think it can).

I'm not trying to argue here, but saying there is a single device that does is this is disingenuous, and stating that other integrations use "the reolink API directly" (or amcrest or whoever) while the onvif core spec supports this is misleading. I understand, and it is completely fine to say you or the Frigate project is not interested in supporting parts of the onvif spec (at this time or ever), but let's call it what it is, not blame one-off devices or vendor specific APIs.

**Reply by NickM-27** [reply]
Thanks for the detailed response. To be clear I don't disagree with what you are saying in general, but I think you misunderstand my point. 

When we support onvif PTZ we support the onvif standard directly and any camera that implements the onvif spec correctly can be added to the config and work.

Conversely, for the onvif events we must connect to the device via their http api to initiate the connection and that means if we implemented this feature for reolink doorbell for example it would not work directly with an amcrest doorbell. The amcrest doorbell would need a separate implementation for that initial logic and then we'd have to hope that they both implement the onvif events similarly after that point.

**Reply by mayo** [reply]
You don't have to guess what the API endpoints are for any capabilities. The device can be inspected the directly, if you know the device IP and onvif port. The IP and port are already defined in Frigate for PTZ (`cameras.[camera_name].onvif.{host,port,user,password}`). From there, the onvif spec dictates that a compliant device needs to have a `/onvif/device_service/` endpoint. One can then inspect the device and learn what capabilities it has. For this particular case, we're interested in Events capability, and what the endpoint for events is.

(edit for clarity) Further, the Events API is well defined as part of the onvif standard, so for a compliant device, the vendor doesn't get to change things up any way they want. The aim of the ONVIF standard is to define a unified interface for interaction and discovery, such that you don't have to have a vendor specific implementation. As long as vendors are compliant, based on the profile and capabilities you can relatively easily discover what the device can do and how.

In both of the vendor examples above (Amcrest and Reolink), their doorbells support ONVIF Profile T, which includes Events and Digital Inputs (among other things), and the doorbell button is exposed as a Digital Input.

For example, asking for Capabilities:

```
curl -s http://<ip>:<port>/onvif/device_service/ -d '<s:Envelope xmlns:s="http://www.w3.org/2003/05/soap-envelope">
<s:Body xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema">
<GetCapabilities xmlns="http://www.onvif.org/ver10/device/wsdl"><Category>All</Category></GetCapabilities>
</s:Body>
</s:Envelope>' | tidy -xml -indent
```
```
<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://www.w3.org/2003/05/soap-envelope"
[...]
    <tds:GetCapabilitiesResponse>
      <tds:Capabilities>
		<tt:Device>
			<tt:XAddr>http://<ip>:<port>/onvif/device_service</tt:XAddr>
			<tt:Network>
[...]
		</tt:Device>
		<tt:Events>
			<tt:XAddr>http://<ip>:<port>/onvif/event_service</tt:XAddr>
			<tt:WSSubscriptionPolicySupport>true</tt:WSSubscriptionPolicySupport>
			<tt:WSPullPointSupport>true</tt:WSPullPointSupport>
			<tt:WSPausableSubscriptionManagerInterfaceSupport>false</tt:WSPausableSubscriptionManagerInterfaceSupport>
		</tt:Events>
[...]
      </tds:Capabilities>
    </tds:GetCapabilitiesResponse>
  </SOAP-ENV:Body>
</SOAP-ENV:Envelope>
```
